From d1d2e97c645805d303fd4bcac3898b80000337da Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@freescale.com>
Date: Fri, 2 Nov 2012 13:51:24 +0800
Subject: [PATCH 18/25] usb: chipidea: imx: select usb id pin using syscon
 interface

otg_id_pin_select_change is set, we need to select usb id pin.

Signed-off-by: Peter Chen <peter.chen@freescale.com>
---
 .../devicetree/bindings/usb/ci13xxx-imx.txt        |    6 +++++
 arch/arm/boot/dts/imx6q-sabrelite.dts              |    1 +
 drivers/usb/chipidea/ci13xxx_imx.c                 |   28 ++++++++++++++++++++
 3 files changed, 35 insertions(+)

diff --git a/Documentation/devicetree/bindings/usb/ci13xxx-imx.txt b/Documentation/devicetree/bindings/usb/ci13xxx-imx.txt
index 493a414..22b4805 100644
--- a/Documentation/devicetree/bindings/usb/ci13xxx-imx.txt
+++ b/Documentation/devicetree/bindings/usb/ci13xxx-imx.txt
@@ -17,6 +17,11 @@ Optional properties:
   that indicate usb controller index
 - vbus-supply: regulator for vbus
 - disable-over-current: disable over current detect
+- otg_id_pin_select_change: for platforms who have otg_id_pin selection
+(Currently, only imx6 has this selection), and need to change
+default value, please add this flag. Below is the pin selection for imx6q:
+	- ENET_RX_ER: 0  (defalt)
+	- GPIO_1: 1
 
 Examples:
 usb@02184000 { /* USB OTG */
@@ -26,4 +31,5 @@ usb@02184000 { /* USB OTG */
 	fsl,usbphy = <&usbphy1>;
 	fsl,usbmisc = <&usbmisc 0>;
 	disable-over-current;
+	otg_id_pin_select_change;
 };
diff --git a/arch/arm/boot/dts/imx6q-sabrelite.dts b/arch/arm/boot/dts/imx6q-sabrelite.dts
index 816428b..7da2a13 100644
--- a/arch/arm/boot/dts/imx6q-sabrelite.dts
+++ b/arch/arm/boot/dts/imx6q-sabrelite.dts
@@ -113,6 +113,7 @@
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_usbotg_1>;
 	disable-over-current;
+	otg_id_pin_select_change;
 	status = "okay";
 };
 
diff --git a/drivers/usb/chipidea/ci13xxx_imx.c b/drivers/usb/chipidea/ci13xxx_imx.c
index ebed129..5a2b196 100644
--- a/drivers/usb/chipidea/ci13xxx_imx.c
+++ b/drivers/usb/chipidea/ci13xxx_imx.c
@@ -22,12 +22,16 @@
 #include <linux/regulator/consumer.h>
 #include <linux/pinctrl/consumer.h>
 #include <linux/usb/of.h>
+#include <linux/regmap.h>
+#include <linux/mfd/syscon.h>
 
 #include "ci.h"
 #include "ci13xxx_imx.h"
 
 #define pdev_to_phy(pdev) \
 	((struct usb_phy *)platform_get_drvdata(pdev))
+#define IOMUXC_IOMUXC_GPR1			0x00000004
+#define USB_OTG_ID_SEL_BIT			(1<<13)
 
 struct ci13xxx_imx_data {
 	struct usb_phy *phy;
@@ -189,6 +193,30 @@ static int ci13xxx_imx_probe(struct platform_device *pdev)
 		}
 	}
 
+#if IS_ENABLED(CONFIG_MFD_SYSCON)
+	/* Any imx6 user who needs to change id select pin, do below */
+	if (of_find_property(pdev->dev.of_node,
+				"otg_id_pin_select_change", NULL)) {
+		struct regmap *iomuxc_gpr;
+		u32 gpr1 = 0;
+
+		iomuxc_gpr = syscon_regmap_lookup_by_compatible
+			("fsl,imx6q-iomuxc-gpr");
+		if (!IS_ERR(iomuxc_gpr)) {
+			/* Select USB ID pin at iomuxc grp1 */
+			regmap_read(iomuxc_gpr, IOMUXC_IOMUXC_GPR1, &gpr1);
+			regmap_write(iomuxc_gpr, IOMUXC_IOMUXC_GPR1,
+					gpr1 | USB_OTG_ID_SEL_BIT);
+		} else {
+			ret = PTR_ERR(iomuxc_gpr);
+			dev_err(&pdev->dev,
+				"failed to find imx6q-iomuxc-gpr regmap:%d\n",
+				ret);
+			goto err_clk;
+		}
+	}
+#endif
+
 	plat_ci = ci13xxx_add_device(&pdev->dev,
 				pdev->resource, pdev->num_resources,
 				pdata);
-- 
1.7.10.4

